electrumsv/wallet.py

```diff
@@ -978,17 +993,31 @@ def get_address_history(self, address: Address):
                 if tx_fee is not None:
                     flags |= TxFlags.HasFee
                 updates.append((tx_hash, data, None, flags))
-            self._datastore.tx.update_or_add(updates)
+
+                self._datastore.tx.update_or_add(updates)
 
             for tx_id in set(t[0] for t in hist):
+                has_tx_data = self._datastore.tx.have_transaction_data(tx_id)
+                is_cached = self._datastore.tx.is_cached(tx_id)
+                if has_tx_data and is_cached:
+                    entry = self._datastore.tx.get_cached_entry(tx_id)
+                if has_tx_data and not is_cached:
+                    entry = self._datastore.tx.get_entry(tx_id)
+
+                # if StateSigned | StateDispatched | StateReceived update --> StateCleared
+                if has_tx_data and entry.flags & (TxFlags.StateCleared | TxFlags.StateSettled) == 0:
+                    self.logger.debug("updating incoming tx --> StateCleared. Before: %s ", entry)
+                    self.set_transaction_state(tx_id, TxFlags.StateCleared | TxFlags.HasByteData)
+                    self.logger.debug("After update: %s ", self._datastore.tx.get_entry(tx_id))
+
                 # if addr is new, we have to recompute txi and txo
-                if self._datastore.tx.have_transaction_data(tx_id) and \
-                        not len(self.get_txins(tx_id, address)) and \
-                        not len(self.get_txouts(tx_id, address)):
-                    tx = self.get_transaction(tx_id)
-                    self.apply_transactions_xputs(tx_id, tx)
+                if len(self.get_txins(tx_id, address)) and len(self.get_txouts(tx_id, address)):
+                    self.apply_transactions_xputs(tx_id, entry.transaction)
```

Comment:
Currently this is not guaranteed to have the transaction, as this only accessed the transaction if the lazily loaded byte data  has been loaded into the cache. See the comment above.